File: coora-app/packages/examples/src/entry/test_memory.ts
// memory.

export function _size(): u32{
	return memory.size()
}
export function _load(ptr: i32): i32{
	// const a = 2
	return load<i32>(ptr)
}
export function _store(ptr: u32, val: i32): void{
	return store<i32>(ptr, val)
}

// export function add(a: i32, b: i32): i32{
// 	return a + b
// }

export function getString(): u32{
	const str = 'hello from wasm'
	return changetype<usize>(str) as u32
}
export function printHello(): void{
	const str = 'hello from wasm'
	println(str)
}

//@ts-ignore externam
@external("Serial", "println")
declare function _println(ptr: usize, len: i32): void;

export function println(str: string): void {
	const out = String.UTF8.encode(str)
	_println(changetype<usize>(out), out.byteLength)
}

export const a: string = '\1\2\3\4'

File: coora-engine/crates/engine/test/wasm/test_memory.rs
use crate::*;
use coora_engine::*;


sweet! {


	it "works" {
		let wasm = include_wasm!("../../../", "test_memory");

		let mut console = StdConsole{}.as_shared();
		let mut app = WasmApp::new();
		// link(&mut app);
		app.add_plugin(&mut console)?;
		app.build_with_wasm(&wasm[..]);

		let mut mem_test = MemoryTestInstance::new(&mut app);
		mem_test.printHello();
		mem_test.printHello();
		mem_test.printHello();

		// let str = str::from_utf8(&a).unwrap();
		// expect(str).to_be("hello from wasm")?;
	}
}

// fn link(app: &mut WasmApp) {
// 	let mut store = app.store.lock().unwrap();
// 	let memory = Arc::clone(&app.memory);
// 	app.linker
// 		.define(
// 			"Serial",
// 			"println",
// 			Func::wrap(
// 				&mut *store,
// 				move |caller: Caller<coora_engine::UserState>, ptr: u32, len: u32| {
// 					let memory = memory.lock().unwrap();
// 					let ctx = caller.as_context();
// 					println!("printing ptr: {ptr}, len: {len}");
// 					let data = &memory.data(ctx)[ptr as usize..(ptr + len) as usize];
// 					let str = str::from_utf8(&data).unwrap();
// 					println!("printing: {str}");

// 					// self3.lock().unwrap().add(a)
// 				},
// 			),
// 		)
// 		.unwrap();
// }
